core/includes/config.inc

```diff
@@ -839,6 +872,303 @@ interface ConfigStorageInterface {
+    }
+    return $data;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function readMultiple(array $names) {
+    $list = array();
+    foreach ($names as $name) {
+      if ($data = $this->read($name)) {
+        $list[$name] = $data;
+      }
+    }
+    return $list;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function write($name, array $data) {
+    // Ensure that the config name is included in the written file.
+    $data = array_merge(array('_config_name' => $name), $data);
+    $data = $this->encode($data) . "\n";
+    $file_path = $this->getFilePath($name);
+    try {
+      db_merge($this->table, array('target' => $this->database))
+        ->key(array('name' => $name))
+        ->fields(array(
+          'name' => $name,
+          'data' => $data,
+          'ctime' => time(),
+        ))
+        ->execute();
+    }
+    catch (\Exception $e) {
+      throw new ConfigStorageException('Failed to write configuration file: ' . $this->getFilePath($name), 0, $e);
+    }
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function delete($name) {
+    if (!$this->exists($name)) {
+      return FALSE;
+    }
+    db_delete($this->table, array('target' => $this->database))
+      ->condition('name', $new_name)
+      ->execute();
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function rename($name, $new_name) {
+    try {
+      db_delete($this->table, array('target' => $this->database))
+        ->condition('name', $new_name)
+        ->execute();
+      db_update($this->table, 'c', array('target' => $this->database))
+        ->fields('c.name', $new_name)
+        ->condition('c.name', $name)
+        ->execute();
+    }
+    catch (\Exception $e) {
+      throw new ConfigStorageException('Failed to rename configuration file from: ' . $this->getFilePath($name) . ' to: ' . $this->getFilePath($new_name), 0, $e);
+    }
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getModifiedTime($name) {
+    $data = db_select($this->table, 'c', array('target' => $this->database))
+      ->fields('c', 'ctime')
+      ->condition('c.name', $name)
+      ->execute()
+      ->fetchField();
+
+    return empty($data) ? FALSE : $data;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function encode($data) {
+    $contents = backdrop_json_encode($data, TRUE);
+    if ($contents === FALSE) {
+      throw new ConfigStorageException(t('The configuration string could not be parsed.'));
+    }
+    return $contents;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function decode($raw) {
+    // Use json_decode() directly for efficiency.
+    $contents = json_decode($raw, TRUE);
+    if (is_null($contents)) {
+      throw new ConfigStorageException('The configuration string could not be parsed.');
+    }
+    return $contents;
+  }
```

Comment:
Initially I thought maybe a base class would make sense here, but with only `encode()` and `decode()` sharing any code (and that is a little dubious, we could serialize instead of using JSON in the future), so I don't think a base class is sensible in this case.